clojure-mode.el

```diff
@@ -1534,6 +1546,197 @@ This will skip over sexps that don't represent objects, so that ^hints and
+
+(defun clojure--ensure-parens-around-function-names ()
+  (unless (looking-at "[\n\r\t ]?(")
+    (skip-syntax-forward " ")
+    (clojure--wrap-around)
+    (backward-sexp)))
+
+(defun clojure--unwind-first ()
+  (forward-sexp)
+  (save-excursion
+    (let ((contents (clojure-delete-and-extract-sexp)))
+      (when (looking-at " *\n")
+        (join-line -1))
+      (clojure--ensure-parens-around-function-names)
+      (down-list)
+      (forward-sexp)
+      (insert contents)))
+  (forward-char))
+
+(defun clojure--pop-out-of-threading ()
+  (down-list 2)
+  (backward-up-list)
+  (raise-sexp)
+  (save-excursion
+    (let ((beg (point))
+          (end (progn
+                 (forward-sexp)
+                 (point))))
+      (clojure-indent-region beg end))))
+
+(defun clojure--nothing-more-to-unwind ()
+  (save-excursion
+    (let ((beg (point)))
+      (forward-sexp)
+      (down-list -1)
+      (backward-sexp 2) ;; the last sexp, the threading macro
+      (when (looking-back "(\\s-*")
+          (backward-up-list)) ;; and the paren
+      (= beg (point)))))
+
+;;;###autoload
+(defun clojure-unwind ()
+  "Unwind thread at point or above point by one level.
+Return nil if there are no more levels to unwind."
+  (interactive)
+  (ignore-errors
+    (forward-char 3))
+  (search-backward-regexp "([^-]*->")
+  (if (clojure--nothing-more-to-unwind)
+      (progn (clojure--pop-out-of-threading)
+             nil)
+    (down-list)
+    (cond
+     ((looking-at "[^-]*->[\n\r\t ]")  (clojure--unwind-first))
+     ((looking-at "[^-]*->>[\n\r\t ]") (clojure--unwind-last)))
+    t))
+
+;;;###autoload
+(defun clojure-unwind-all ()
+  "Fully unwind thread at point or above point."
+  (interactive)
+  (while (clojure-unwind)
+    t))
+
+(defun clojure--remove-superfluous-parens ()
+  (when (looking-at "([^ )]+)")
+    (down-list)
+    (raise-sexp)))
+
+(defun clojure--thread-first ()
+  (down-list)
+  (forward-symbol 1)
+  (let* ((beg (point))
+         (end (progn
+                (if (looking-at ")")
+                    (up-list)
+                  (forward-list))
+                (point)))
+         (contents (buffer-substring beg end)))
+    (if (string= contents ")")
+        nil
+      (delete-region beg end)
+      (backward-up-list)
+      (just-one-space 0)
+      (insert contents)
+      (newline-and-indent)
+      (clojure--remove-superfluous-parens)
+      t)))
+
+(defun clojure--thread-last ()
+  (forward-sexp 2)
+  (down-list -1)
+  (let* ((end (point))
+         (beg (progn (backward-sexp)
+                     (point)))
+         (contents (buffer-substring beg end)))
+    (if (looking-back "(")
+        nil
+      (delete-region beg end)
+      (just-one-space 0)
+      (backward-up-list)
+      (insert contents)
+      (newline-and-indent)
+      (clojure--remove-superfluous-parens)
+      ;; cljr #255 Fix dangling parens
+      (backward-up-list)
+      (forward-sexp)
+      (when (looking-back "^\\s-*)+\\s-*")
+        (join-line))
```

Comment:
Not sure if that can happen, but this will break the code if the line above ends in a comment.